lib/scholar/cluster/hierarchical.ex

```diff
@@ -0,0 +1,476 @@
+            i <- 0..(Nx.size(links) - 1) do
+        # i < j because of how links is formed.
+        # i will become the new clade index and we "infinity-out" j.
+        j = links[i]
+
+        if j == n do
+          {clades, count, pointers, pairwise, diss, sizes, links}
+        else
+          # Clades a and b (i and j of pairwise) are being merged into c.
+          a = find_clade(pointers, i)
+          b = find_clade(pointers, j)
+          c = count + n
+
+          # Update clades
+          new_clade = Nx.stack([a, b]) |> Nx.sort() |> Nx.new_axis(0)
+          clades = Nx.put_slice(clades, [count, 0], new_clade)
+
+          # Update sizes
+          sc = sizes[i] + sizes[j]
+          sizes = Nx.indexed_put(sizes, Nx.stack([i, c]) |> Nx.new_axis(-1), Nx.stack([sc, sc]))
+
+          # Update dissimilarities
+          diss = Nx.indexed_put(diss, Nx.stack([count]), pairwise[i][j])
+
+          # Update pairwise
+          {pairwise, _, _, _, _, _} =
+            while {pairwise, x = i, y = j, sa = sizes[i], sb = sizes[j], sc = sc},
+                  z <- 0..(n - 1) do
+              if z == x or z == y or Nx.is_infinity(pairwise[[0, z]]) do
+                {pairwise, x, y, sa, sb, sc}
+              else
+                dac = pairwise[[x, z]]
+                dbc = pairwise[[y, z]]
+                dab = pairwise[[x, y]]
+                update = update_fun.(dac, dbc, dab, sa, sb, sc)
+
+                # TODO: do this in a single call?
+                pairwise =
+                  pairwise
+                  |> Nx.indexed_put(Nx.stack([x, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, x]), update)
+                  |> Nx.indexed_put(Nx.stack([y, z]), update)
+                  |> Nx.indexed_put(Nx.stack([z, y]), update)
+
+                {pairwise, x, y, sa, sb, sc}
+              end
+            end
+
+          infinities = Nx.take_diagonal(pairwise)
+
+          pairwise =
+            pairwise
+            |> Nx.put_slice([j, 0], Nx.reshape(infinities, {1, n}))
+            |> Nx.put_slice([0, j], Nx.reshape(infinities, {n, 1}))
+
+          # Update pointers
+          indices = [i, j] |> Nx.stack() |> Nx.new_axis(-1)
+          pointers = Nx.indexed_put(pointers, indices, Nx.stack([c, c]))
+
+          {clades, count + 1, pointers, pairwise, diss, sizes, links}
+        end
+      end
+
+    {clades, count, pointers, pairwise, diss, sizes}
+  end
+
+  defnp find_clade(pointers, i) do
+    {i, _, _} =
+      while {_current = i, next = pointers[i], pointers}, next != -1 do
+        {next, pointers[next], pointers}
+      end
+
+    i
+  end
+
+  # Dissimilarity functions
+
+  defnp pairwise_euclidean(%Nx.Tensor{} = x) do
+    x |> pairwise_euclidean_sq() |> Nx.sqrt()
+  end
+
+  defnp pairwise_euclidean_sq(%Nx.Tensor{} = x) do
+    sq = Nx.sum(x ** 2, axes: [1], keep_axes: true)
+    sq + Nx.transpose(sq) - 2 * Nx.dot(x, [1], x, [1])
+  end
```

Comment:
These functions have been already implemented in Scholar ;)

Reply:
Oh where? I didn't see a public function for them.

Reply:
Here https://github.com/elixir-nx/scholar/blob/dacf106f63125a3b57866e19449fdf15cc7a27ca/lib/scholar/metrics/distance.ex#L553-L651

Reply:
Oh so it is! Thanks!

I was thrown off because it's not on hexdocs:

https://hexdocs.pm/scholar/Scholar.Metrics.Distance.html

Reply:
[ad9ddb4](https://github.com/elixir-nx/scholar/pull/187/commits/ad9ddb47f77c15e129ed67a5f89f75b43a89508f)